### PR TITLE
[Ready for Testmerge] Implement Diagonal Movement

### PR DESCRIPTION
**Pull request by Couls**
**What does this PR do:**
Port over TG diagonal movement and their keybinding system(macros will not work for diagonal movement)
ezgif-5-66705f348b05
ezgif-5-5675aa4e8e93

**Changelog:**
:cl:
add: Diagonal movement
add: Input subsystem(numpad targetting, press numpad 8 multiple times to target eyes and mouth, numpad 6 or 4 to target arms and press them again to target hands and numpad 1 or 3 to target legs and press them again to target feet) taken from https://github.com/tgstation/tgstation/pull/32751
add: change confused status to have you move diagonally randomly in the direction you're headed if not too confused(now you can drunkenly walk down the hall)
add: clients are now children of datums like everything else in BYOND taken from https://github.com/tgstation/tgstation/pull/20394
add: AZERTY and numpad targetting preferences
/:cl:

- [x] Diagonal movement
- [x] Port the keybinding system
- [x] Fix keybinding system
- [x] Actually take inputs and it'll be ready for review
- [x] Fix quick equip to belt and bag(it currently says you have no backpack when you do)
- [x] Implement AZERTY support (numpad targetting toggle is in too)
- [x] Fix freezing caused by neckgrabs and resting no longer works too[probably mob_movement
- [x] fix numpad targetting not working

**NOW REQUIRES SQL UPDATE**

### CODE REVIEW

**Comment by KasparoVy** [comment]
Holy fuck, you can hotkey selecting specific zones on the body now.

This is going to be mega robust

File: code/modules/mob/mob_movement.dm [hunk 1]
```diff
@@ -461,3 +490,123 @@
...
+			next_in_line = BODY_ZONE_PRECISE_EYES
+		if(BODY_ZONE_PRECISE_EYES)
+			next_in_line = BODY_ZONE_PRECISE_MOUTH
+		else
+			next_in_line = BODY_ZONE_HEAD
+
+	var/obj/screen/zone_sel/selector = mob.hud_used.zone_select
+	selector.set_selected_zone(next_in_line, mob)
+
+/client/verb/body_r_arm()
```
Comment (KasparoVy) [inline code/modules/mob/mob_movement.dm]: Could we implement procs to select body_r_hand, body_l_hand, body_r_foot & body_l_foot too?

File: code/modules/mob/mob_movement.dm [hunk 2]
```diff
@@ -461,3 +490,123 @@
...
+	set name = "body-r-arm"
+	set hidden = 1
+	if(!check_has_body_select())
+		return
+
+	var/next_in_line
+	if (mob.zone_selected == BODY_ZONE_R_ARM)
```
Comment (KasparoVy) [inline code/modules/mob/mob_movement.dm]: No spessing b/w `if` and `()`

File: code/modules/mob/mob_movement.dm [hunk 3]
```diff
@@ -461,3 +490,123 @@
...
+
+/client/verb/body_l_arm()
+	set name = "body-l-arm"
+	set hidden = 1
+	
+	if(!check_has_body_select())
+		return
+		
+	var/next_in_line
+	if (mob.zone_selected == BODY_ZONE_L_ARM)
```
Comment (KasparoVy) [inline code/modules/mob/mob_movement.dm]: Spessing

File: code/modules/mob/mob_movement.dm [hunk 4]
```diff
@@ -461,3 +490,123 @@
...
+
+/client/verb/body_r_leg()
+	set name = "body-r-leg"
+	set hidden = 1
+
+	if(!check_has_body_select())
+		return
+		
+	var/next_in_line
+	if (mob.zone_selected == BODY_ZONE_R_LEG)
```
Comment (KasparoVy) [inline code/modules/mob/mob_movement.dm]: Spessing

File: code/modules/mob/mob_movement.dm [hunk 5]
```diff
@@ -461,3 +490,123 @@
...
+
+/client/verb/body_l_leg()
+	set name = "body-l-leg"
+	set hidden = 1
+
+	if(!check_has_body_select())
+		return
+		
+	var/next_in_line
+	if (mob.zone_selected == BODY_ZONE_L_LEG)
```
Comment (KasparoVy) [inline code/modules/mob/mob_movement.dm]: Spessing

**Comment by SkeletalElite** [comment]
> Holy fuck, you can hotkey selecting specific zones on the body now (nvm already mapped to numpad, but 'can map them to something more convenient' heheh. Only caviat is that to aim for hands you have to hit the arm select hotkey twice, or to aim for eyes you gotta hit the head hotkey twice.
> 
> This is going to be mega robust, I've been thinking about that for ages! This is the future!

Yeah its going to be a lot better than what we have now but Id still like to see a radial menu for body part targeting

File: interface/skin.dmf [hunk 6]
```diff
@@ -1484,108 +114,97 @@ menu "menu"
...
+		anchor2 = none
+		background-color = none
 		is-default = true
-		is-maximized = true
+		saved-params = "pos;size;is-minimized;is-maximized"
 		title = "Paradise Station 13"
+		is-maximized = true
 		icon = 'icons\\paradise.png'
-		menu = "menu"
 		macro = "macro"
```
Comment (Markolie) [inline interface/skin.dmf]: This needs to be set to "default".

File: code/modules/keybindings/setup.dm [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+		if(setname != "default")
+			winclone(src, "default", setname)
+		var/list/macro_set = macro_sets[setname]
+		for(var/k in 1 to macro_set.len)
+			var/key = macro_set[k]
+			var/command = macro_set[key]
+			winset(src, "[setname]-[REF(key)]", "parent=[setname];name=[key];command=[command]")
+
+	if(prefs.toggles & HOTKEYS)
+		winset(src, null, "mapwindow.map.focus=true input.background-color=#e0e0e0")
```
Comment (Markolie) [inline code/modules/keybindings/setup.dm]: This doesn't set hotkeys yet.

File: code/__DEFINES/combat.dm [hunk 8]
```diff
@@ -107,10 +107,6 @@
 #define BODY_ZONE_PRECISE_EYES		"eyes"
 #define BODY_ZONE_PRECISE_MOUTH		"mouth"
 #define BODY_ZONE_PRECISE_GROIN		"groin"
-#define BODY_ZONE_PRECISE_L_HAND	"l_hand"
```
Comment (KasparoVy) [inline code/__DEFINES/combat.dm]: Uh oh...

Comment (Couls) [reply code/__DEFINES/combat.dm]: WHOOPS

File: code/modules/keybindings/setup.dm (hunk 7 above)
Comment (Couls) [reply code/modules/keybindings/setup.dm]: I still don't know what to do with this yet honestly

**Comment by AffectedArc07** [comment]
Can we not have the numpad region targeting? This broke my numpad full of emotes before, and a **lot** of people also have numpads full of emotes, who will also be pissed

**Comment by Couls** [comment]
Once this is working I’ll make it an config option along with AZERTY support

**Comment by KasparoVy** [comment]
Could always have the numpad emotes require another key be pressed *shrug*

I have mine all long the standard number row but you need to use shift too.
You can also make use of newlines in the macros to have multiple emotes for different races bound to the same key, I. E. Vox rattle, vulp growl + unathi hiss all on one macro

**Comment by Couls** [comment]
I apologize in advance to the poor maintainer who will have to review this frankenstein abomination of code

File: code/__DEFINES/colors.dm [hunk 9]
```diff
@@ -0,0 +1,54 @@
+
+#define COLOR_INPUT_DISABLED "#F0F0F0"
```
Comment (AffectedArc07) [inline code/__DEFINES/colors.dm]: But why

File: code/__DEFINES/preferences.dm [hunk 10]
```diff
@@ -29,7 +29,6 @@
 #define UI_DARKMODE 131072
 #define DISABLE_KARMA 262144
 #define CHAT_NO_MENTORTICKETLOGS 524288
-
```
Comment (AffectedArc07) [inline code/__DEFINES/preferences.dm]: What did you even do here

File: code/__DEFINES/flags.dm [hunk 11]
```diff
@@ -1,6 +1,11 @@
 #define ALL ~0 //For convenience.
 #define NONE 0
 
+// for /datum/var/datum_flags
+#define DF_USE_TAG		(1<<0)
+#define DF_VAR_EDITED	(1<<1)
+#define DF_ISPROCESSING (1<<2)
+
```
Comment (AffectedArc07) [inline code/__DEFINES/flags.dm]: Why was this added?

File: code/modules/client/preference/preferences_mysql.dm [hunk 12]
```diff
@@ -55,7 +55,7 @@
 	ooccolor		= sanitize_hexcolor(ooccolor, initial(ooccolor))
 	UI_style		= sanitize_inlist(UI_style, list("White", "Midnight"), initial(UI_style))
 	default_slot	= sanitize_integer(default_slot, 1, max_save_slots, initial(default_slot))
-	toggles			= sanitize_integer(toggles, 0, 2097151, initial(toggles))
+	toggles			= sanitize_integer(toggles, 0, 4194304, initial(toggles))
```
Comment (AffectedArc07) [inline code/modules/client/preference/preferences_mysql.dm]: Why did you increase this of you didnt add a toggle?
And if you did it should be `4194303`

File: interface/interface.dm [hunk 13]
```diff
@@ -144,56 +145,60 @@ Admin ghost:
...
-\ta = Move Left
-\ts = Move Down
-\td = Move Right
-\tw = Move Up
-\tq = Drop Item
-\te = Equip Item
-\tr = Throw Item
-\tm = Me
-\tt = Say
+\tTAB = toggle hotkey-mode
```
Comment (AffectedArc07) [inline interface/interface.dm]: Why lowercase all of this?

File: interface/interface.dm [hunk 14]
```diff
@@ -144,56 +145,60 @@ Admin ghost:
...
-\tx = Swap Hands
-\tz = Activate Held Object (or y)
-\tf = Cycle Intents Left
-\tg = Cycle Intents Right
-\t1 = Help Intent
-\t2 = Disarm Intent
-\t3 = Grab Intent
-\t4 = Harm Intent
+\tb = resist
+\t<B></B>h = stop pulling
```
Comment (AffectedArc07) [inline interface/interface.dm]: Why bold and unbold?

File: interface/interface.dm [hunk 15]
```diff
@@ -144,56 +145,60 @@ Admin ghost:
...
+\tx = swap-hand
+\tz = activate held object (or y)
```
Comment (AffectedArc07) [inline interface/interface.dm]: z/y = activate held object

File: interface/interface.dm [hunk 16]
```diff
@@ -144,56 +145,60 @@ Admin ghost:
...
-\tINS = Cycle Intents Right
-\tHOME = Drop Item
-\tPGUP = Swap Hands
-\tPGDN = Activate Held Object
-\tEND = Throw Item
-\tF2 = OOC
-\tF3 = Say
-\tF4 = Me
+\tCtrl+x = swap-hand
+\tCtrl+z = activate held object (or Ctrl+y)
```
Comment (AffectedArc07) [inline interface/interface.dm]: Same here

File: code/__DEFINES/flags.dm (hunk 11 above)
Comment (Couls) [reply code/__DEFINES/flags.dm]: its used in REF

File: code/__DEFINES/colors.dm (hunk 9 above)
Comment (Couls) [reply code/__DEFINES/colors.dm]: taken from TG, just helps readability its the color of the input bar when you're in hotkey mode

File: code/__DEFINES/preferences.dm (hunk 10 above)
Comment (Couls) [reply code/__DEFINES/preferences.dm]: this is outdated, I added in two preferences for AZERTY mode and numpad targetting since

**Comment by Couls** [comment]
There's a bug where you have to reconnect before the inputs actually work and I have no idea whats causing this if anyone reviewing can figure it out that would be great

File: code/modules/client/client procs.dm [hunk 17]
```diff
@@ -400,9 +400,10 @@
 
 	if(!winexists(src, "asset_cache_browser")) // The client is using a custom skin, tell them.
 		to_chat(src, "<span class='warning'>Unable to access asset cache browser, if you are using a custom skin file, please allow DS to download the updated version, if you are not, then make a bug report. This is not a critical issue but can cause issues with resource downloading, as it is impossible to know when extra resources arrived to you.</span>")
-
-	if(SSinput.initialized)
-		set_macros()
+	
+	if (SSinput.initialized)
+		spawn(30) // without this input won't work on first connect
```
Comment (Couls) [inline code/modules/client/client procs.dm]: This line fixed the issue

File: code/modules/client/client procs.dm [hunk 18]
```diff
@@ -400,9 +400,10 @@
 
 	if(!winexists(src, "asset_cache_browser")) // The client is using a custom skin, tell them.
 		to_chat(src, "<span class='warning'>Unable to access asset cache browser, if you are using a custom skin file, please allow DS to download the updated version, if you are not, then make a bug report. This is not a critical issue but can cause issues with resource downloading, as it is impossible to know when extra resources arrived to you.</span>")
-
-	if(SSinput.initialized)
-		set_macros()
+	
+	if (SSinput.initialized)
```
Comment (KasparoVy) [inline code/modules/client/client procs.dm]: Spessing :rofl:

File: code/modules/client/client procs.dm (hunk 17 above)
Comment (KasparoVy) [reply code/modules/client/client procs.dm]: That's wack, I wonder why it has to chill for so long. Maybe its init order should be bumped down instead? Hm...

File: code/__HELPERS/unsorted.dm [hunk 19]
```diff
@@ -2029,3 +2029,17 @@ var/mob/dview/dview_mob = new
 
 /proc/pass()
 	return
+	
+// \ref behaviour got changed in 512 so this is necesary to replicate old behaviour.
+// If it ever becomes necesary to get a more performant REF(), this lies here in wait
+// #define REF(thing) (thing && istype(thing, /datum) && (thing:datum_flags & DF_USE_TAG) && thing:tag ? "[thing:tag]" : "\ref[thing]")
+/proc/REF(input)
```
Comment (Fox-McCloud) [inline code/__HELPERS/unsorted.dm]: be use `UID` instead of `REF`

File: code/__DEFINES/flags.dm [hunk 20]
```diff
@@ -1,6 +1,11 @@
 #define ALL ~0 //For convenience.
 #define NONE 0
 
+// for /datum/var/datum_flags
+#define DF_USE_TAG		(1<<0)
+#define DF_VAR_EDITED	(1<<1)
+#define DF_ISPROCESSING (1<<2)
```
Comment (Fox-McCloud) [inline code/__DEFINES/flags.dm]: the above two aren't used.

File: code/__HELPERS/unique_ids.dm [hunk 21]
```diff
@@ -26,7 +26,7 @@ var/global/next_unique_datum_id = 1
 		tag = tag_backup
 	return unique_datum_id
 
-/client/proc/UID()
+/client/UID()
```
Comment (Fox-McCloud) [inline code/__HELPERS/unique_ids.dm]: don't override this completely; just get rid of it, otherwise, there's going to be problems.

File: code/modules/mob/dead/dead.dm [hunk 22]
```diff
@@ -19,5 +19,6 @@
 
 	for(var/datum/light_source/L in light_sources)
 		L.source_atom.update_light()
-
 	return 1
+/mob/dead/Initialize()
```
Comment (Fox-McCloud) [inline code/modules/mob/dead/dead.dm]: `mapload`

**Comment by Mitchs98** [comment]
Fix freezing caused by neckgrabs and resting no longer works too

What does this mean? You can't rest anymore? Or? If so, that's...definitely shouldn't be a thing. This needs to be more clear on what it does.

**Comment by Couls** [comment]
There was an issue where you couldn’t move when you had someone in a neck grab and you also couldn’t rest(the verb just wouldn’t work for a bit) both have been fixed hence the check mark

**Comment by Mitchs98** [comment]
Ohhh. Okay. Sorry. I thought that might be it but I get easily confused. Just thought I'd be sure ^-^'

File: code/__HELPERS/unsorted.dm (hunk 19 above)
Comment (Couls) [reply code/__HELPERS/unsorted.dm]: Moved the code for this under UID can you make sure what I did won't break anything? I tested it for a bit and it didn't seem to break anything major

File: code/__HELPERS/unique_ids.dm [hunk 23]
```diff
@@ -16,22 +16,26 @@
 
 var/global/next_unique_datum_id = 1
 
-/client/var/tmp/unique_datum_id = null
-
-/datum/proc/UID()
-	if(!unique_datum_id)
+// /client/var/tmp/unique_datum_id = null
+
+/datum/proc/UID(input)
```
Comment (Fox-McCloud) [inline code/__HELPERS/unique_ids.dm]: I wasn't suggesting making `UID` function like `REF`. That'll probably break things in a horrifying manner.

I was suggesting making whatever TG has work with `UID` and not `REF`.

Comment (Couls) [reply code/__HELPERS/unique_ids.dm]: yeah thats kinda what I did, when theres no argument passed as input it just uses UID like it does normally but if there is an argument passed to UID it'll treat it like TG treats REF

**Comment by MrMagolor** [comment]
>tg diagonal movement

That's a funny way to spell goonstation

Also, as for limb targeting: I was gonna say use the goon regular number hotkeys for targeting, but then I realized that goon lacks hands/feet/eyes/mouth/lower torso targeting.

**Comment by Couls** [comment]
> > tg diagonal movement
> 
> That's a funny way to spell goonstation

it's TG's diagonal movement because I took the code from it from TG and I also took the input subsystem from their code so..

**Comment by Couls** [comment]
Holding off on resolving conflicts until I know what maints want to do with thid

**Comment by MrMagolor** [comment]
I think with Lavaland, this is pretty necessary with dodging megafauna attacks.

**Comment by KasparoVy** [comment]
Dancing with dragons is pretty bonkers with a crusher and no diag, but it's just barely workable if you somehow manage to be perfect. Diagonal movement would be glorious

**Comment by Couls** [comment]
Been open for a month now, an update would be nice

**Comment by Spacemanspark** [comment]
Might want to resolve the conflicts so you're waiting less when the maints get around to looking at this.

**Comment by Couls** [comment]
Conflicts Resolved.

File: code/game/atoms_movable.dm [hunk 24]
```diff
@@ -133,40 +133,59 @@
...
 			moving_diagonally = FIRST_DIAG_STEP
-			if(direct & 1)
-				if(direct & 4)
-					if(step(src, NORTH))
+			var/first_step_dir
+			// The `&& moving_diagonally` checks are so that a forceMove taking
+			// place due to a Crossed, Bumped, etc. call will interrupt
+			// the second half of the diagonal movement, or the second attempt
+			// at a first half if step() fails because we hit something.
+			if (direct & NORTH)
```
Comment (variableundefined) [inline code/game/atoms_movable.dm]: No space between if and ( in this file please

File: code/modules/mob/mob_movement.dm [hunk 25]
```diff
@@ -17,8 +17,8 @@
 
 //The byond version of these verbs wait for the next tick before acting.
 //	instant verbs however can run mid tick or even during the time between ticks.
-#define DO_MOVE(this_dir) var/final_dir = turn(this_dir, -dir2angle(dir)); Move(get_step(mob, final_dir), final_dir);
-
+//#define DO_MOVE(this_dir) var/final_dir = turn(this_dir, -dir2angle(dir)); Move(get_step(mob, final_dir), final_dir);
```
Comment (variableundefined) [inline code/modules/mob/mob_movement.dm]: Are there any reason to keep these commented instead of deleted?

**Comment by variableundefined** [comment]
During the test merge it seems this broke movement for some people inconsistently. Problems reported included:
1. Arrow keys being assigned to chat bar color toggle
2. A small menu is brought up to the bottom left of the screen and color is changed when hotkey is activated.
3. People cannot move at all with tab or not.

I cannot replicate it myself locally for now.

File: code/modules/mob/mob_movement.dm (hunk 25 above)
Comment (Couls) [reply code/modules/mob/mob_movement.dm]: Nope, this is a byproduct of testing so I could easily go back to the old system

File: code/modules/client/client procs.dm (hunk 17 above)
Comment (Couls) [reply code/modules/client/client procs.dm]: @variableundefined perhaps we could think of a better way to deal with this? The problem was fixed locally by doing this

**Comment by Couls** [comment]
Holy crap ninjanomnom fixed it

File: code/modules/spacepods/spacepod.dm [hunk 26]
```diff
@@ -1037,7 +1037,7 @@ obj/spacepod/proc/add_equipment(mob/user, var/obj/item/spacepod_equipment/SPE, v
 		return stop()
 
 /obj/spacepod/relaymove(mob/user, direction)
-	if(usr != src.pilot)
+	if(user != src.pilot)
```
Comment (Couls) [inline code/modules/spacepods/spacepod.dm]: *slow clap*

**Comment by KasparoVy** [comment]
God bless
